kotlin/perform-valve-isolation-trace/src/main/java/com/esri/arcgisruntime/sample/performvalveisolationtrace/MainActivity.kt

```diff
@@ -0,0 +1,295 @@
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceType
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.android.synthetic.main.spinner_text_item.view.*
+import java.util.UUID
+
+class MainActivity : AppCompatActivity() {
+
+  private val TAG = MainActivity::class.java.simpleName
+
+  // create a graphics overlay for the starting location and add it to the map view
+  private val startingLocationGraphicsOverlay: GraphicsOverlay by lazy { GraphicsOverlay() }
+
+  // create and apply renderers for the starting point graphics overlay
+  private val startingPointSymbol: SimpleMarkerSymbol by lazy {
+    SimpleMarkerSymbol(
+      SimpleMarkerSymbol.Style.CROSS,
+      Color.GREEN,
+      25f
+    )
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // load the utility network data from the feature service and create feature layers
+    val distributionLineFeatureTable =
+      ServiceFeatureTable(getString(R.string.distribution_line_url))
+    val distributionLineLayer = FeatureLayer(distributionLineFeatureTable)
+    val deviceFeatureTable =
+      ServiceFeatureTable(getString(R.string.device_url))
+    val deviceLayer = FeatureLayer(deviceFeatureTable)
+
+    // create a map with the utility network distribution line and device layers
+    val map = ArcGISMap(Basemap.createStreetsNightVector()).apply {
+      // add the feature layers to the map
+      operationalLayers.addAll(listOf(distributionLineLayer, deviceLayer))
+      // create and load the utility network
+      addDoneLoadingListener {
+        createUtilityNetwork()
+      }
+    }
+
+    mapView.apply {
+      this.map = map
+      // set the starting location graphic overlay's renderer and add it to the map view
+      startingLocationGraphicsOverlay.renderer = SimpleRenderer(startingPointSymbol)
+      graphicsOverlays.add(startingLocationGraphicsOverlay)
+
+      // make sure the fab doesn't obscure the attribution bar
+      addAttributionViewLayoutChangeListener { _, _, _, _, bottom, _, _, _, oldBottom ->
+        val layoutParams = fab.layoutParams as CoordinatorLayout.LayoutParams
+        layoutParams.bottomMargin += bottom - oldBottom
+      }
+      // close the options sheet when the map is tapped
+      onTouchListener = object : DefaultMapViewOnTouchListener(this@MainActivity, mapView) {
+        override fun onTouch(view: View?, event: MotionEvent?): Boolean {
+          if (fab.isExpanded) {
+            fab.isExpanded = false
+          }
+          return super.onTouch(view, event)
+        }
+      }
+    }
+    // show the options sheet when the floating action button is clicked
+    fab.setOnClickListener {
+      fab.isExpanded = !fab.isExpanded
+    }
+  }
+
+  /**
+   * Creates a utility network from the string resource url and loads it.
+   * Initializes the starting point and UI.
```

Comment:
this method descriptions makes it seem like it's handling two different things and therefore goes against the grain of OOP (abstraction): could we extract the UI logic into its method and call that method in here instead? That would also allow our customers to focus on our API call without the bulky UI logic being mixed up in there.

Reply:
I see what you mean, though the resulting function has a lot of parameters because of the button onclick calling performTrace(...):

```kotlin
private fun setupUI(
    networkDefinition: UtilityNetworkDefinition,
    utilityNetwork: UtilityNetwork,
    traceConfiguration: UtilityTraceConfiguration,
    startingLocation: UtilityElement
  ) {...}
```

I could move just the `traceButton.setOnClickListener{}` back into the previous function, or I could make these class level variables, but neither feels very satisfying. What do you think?

Reply:
It looks like you should be able to just pass in the utilityNetworkDefinition with the following, then call that method in `createUtilityNetwork` where the original code was: I think this should work though haven't tested rigorously. What do you think, will this work?

```
  /**
   * Initialize the UI using a utility network definition.
   * 
   * @param networkDefinition the utility network definition to populate the spinner
   */
  private fun setupUI(
    networkDefinition: UtilityNetworkDefinition
  ) {
    // populate the spinner with utility categories as the data and their names as the text
    spinner.adapter =
  object : ArrayAdapter<UtilityCategory>(
    this,
    R.layout.spinner_text_item,
    networkDefinition.categories
  ) {
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
      val spinnerItem = LayoutInflater.from(this@MainActivity)
        .inflate(R.layout.spinner_text_item, parent, false)
      spinnerItem.textView.text = (getItem(position) as UtilityCategory).name
      return spinnerItem
    }

    override fun getDropDownView(
      position: Int,
      convertView: View?,
      parent: ViewGroup
    ): View = getView(position, convertView, parent)
  }
}
```

Reply:
yeah this works, just leaves a single line of ui stuff in the other function! I'll put this in.